test/cpp/end2end/client_lb_end2end_test.cc

```diff
@@ -153,7 +153,6 @@ class ClientLbEnd2endTest : public ::testing::Test {
     for (size_t i = 0; i < servers_.size(); ++i) {
       servers_[i]->Shutdown();
     }
-    grpc_shutdown();
```

Comment:
Wouldn't it be simpler to just change this to use `grpc_shutdown_blocking()`?

Reply:
That would not work because the test class has some members such as server/channel/creds which all carry grpc_init/shutdown in their life time... The alternative I thought about is to add a clear method to the test class, which destroys all of those objects (and wait for grpc_shutdown to terminate). But I feel like that is a bit more fragile because someone may add a new object later but forgot to update the Clear method, and the test becomes flaky again for no clear reason...

Another alternative Juanli proposed is for the next grpc_init wait for the on-going shutdown to finish before starting to init again. I think that could solve this issue but does not sound like the right thing to do. Since in general if we know we are going to init again there is no point to do a shutdown first.

As I mentioned in the desc, I think it is likely a mistake to tie the lifetime of a registered lb policy factory to grpc_init/shutdown anyway. That is another reason I chose this approach.

Sorry for the long reply...

Reply:
So, I actually disagree that it's a mistake to tie the lifetime of a registered LB policy factory to grpc_init/shutdown.  We have tests (like this one) where we register something that's intended only for use in that one test, and we don't want it to be visible outside of the individual test.  Also, since most LB policies will be registered via a plugin init method, and plugin init methods run from grpc_init(), failing to de-register at shutdown would cause conflicts.

I'm really not keen on requiring a static member of the class to point the policy at the right test instance.  This is effectively requiring global state, which seems fairly messy.

Would it work to call `grpc_shutdown_blocking()` if we do it in the test class's dtor instead of in the `TearDown()` method?

Reply:
I do not think that will work because the member objects need to be destructed first. If you prefer, I can add a clear() method instead.

I think whether to link the registry lifetime to grpc_init/shutdown depends on the actual user-facing API to register a policy. If the API is something like registering a plugin, then we are fine, though it may not be very efficient.

Reply:
How about just having the `TearDown()` method explicitly clear out any other data members of the test class before it calls `grpc_shutdown_blocking()`?

Reply:
Sure. That is pretty much the Clear() method I mentioned. I will try to do that.